### PR TITLE
[v2 refactoring proposal] Decouple the runner creation from v1 schema

### LINKED ISSUE #5797: Explore removing `latestv1.Pipeline` and `latestv1.Artifact` from runner & runcontexts everywhere.
Follow up from #5781 PRs

Support Skaffold cmd parser to have multiple pipeline runners.

We are adding a new render phase in a v2 runner. 
This runner should be created when a V3 config is detected. However this is not possible right now. 

The runner interface has  `latestV1` config embedded in it.
```
type Runner interface {
	Apply(context.Context, io.Writer) error
	ApplyDefaultRepo(tag string) (string, error)
	Build(context.Context, io.Writer, []*latestV1.Artifact) ([]graph.Artifact, error)
	Cleanup(context.Context, io.Writer) error
	Dev(context.Context, io.Writer, []*latestV1.Artifact) error
	Deploy(context.Context, io.Writer, []graph.Artifact) error
	DeployAndLog(context.Context, io.Writer, []graph.Artifact) error
	GeneratePipeline(context.Context, io.Writer, []*latestV1.SkaffoldConfig, []string, string) error
	HasBuilt() bool
	HasDeployed() bool
	Prune(context.Context, io.Writer) error
	Render(context.Context, io.Writer, []graph.Artifact, bool, string) error
	Test(context.Context, io.Writer, []graph.Artifact) error
}
```

Similarly see [RunContext](https://github.com/GoogleContainerTools/skaffold/blob/f08c6d68c405d77c6b7f54e56d2951af1bf4f6ca/pkg/skaffold/runner/runcontext/context.go#L37) has `latestv1.Pipeline` embedded in the struct.

The current [V2 runner](https://github.com/GoogleContainerTools/skaffold/tree/master/pkg/skaffold/runner/v2) has the following methods
```

Dev(ctx context.Context, out io.Writer, artifacts []*latestV1.Artifact) error
GeneratePipeline(ctx context.Context, out io.Writer, configs []*latestV1.SkaffoldConfig, configPaths []string, fileOut string) error
Render(ctx context.Context, out io.Writer, builds []graph.Artifact, offline bool, filepath string) 
```
where  `[]graph.Artifact` has `latestV1.Artifact` field.

However, the V2 runner is expected to see a `latestV2.Artifact` and `latestV2.SkaffoldConfig`.  If we make that change, the v2 runner will not adhere by `runner.Runner` interface. 
This will mean, we will have to duplicate all code in cmd package e.g. `cmd.createNewRunner` to return a v2 runner.

It would be great 
1) if we can keep the code in `cmd` package as is
2)  have less duplication in `pkg/runner.Build` functions as build is going to be the same.

Device a plan to remove the version embedded in Runner interface, RunContext struct.

### PR DESCRIPTION
**Pull request by yuwenma**
**Related**: #5673 

Fixes #5797 

**Description**
In runner creation, the schema is no longer bundled with v1 schema (latestV1.skaffoldConfig) but used the interface util.VersionedConfig. 

All v1 commands are manually tested to make sure this PR does not break existing commands.

**Follow-up Work (remove if N/A)**
Distinguish v1 and v2 schema based on the in-use apiVersion in skaffold.yaml.

### CODE REVIEW

**Comment by yuwenma** [comment]
@nkubala gentle ping.

**Comment by yuwenma** [comment]
rebased and merge conflict resolved.

File: cmd/skaffold/app/cmd/runner.go
Comment (nkubala) [inline cmd/skaffold/app/cmd/runner.go:87]: should this be addressed in this PR?

**Review by nkubala — COMMENTED** [review]
I see a few `TODO` leftover in this PR, but I'm not sure they match up with the follow up work detailed in the description. if those aren't going to be addressed in this PR, can we leave a short note in the PR description for what needs to be done later?

I'm also feeling a bit strange about moving the type checking from the method signature into a simple typecast in the method - this seems brittle to me. as an example:

```golang
func targetArtifacts(opts config.SkaffoldOptions, configs []util.VersionedConfig) []*latestV1.Artifact {
	var targetArtifacts []*latestV1.Artifact
	for _, cfg := range configs {
		for _, artifact := range cfg.(*latestV1.SkaffoldConfig).Build.Artifacts {
```

it might be better to abstract this away into `Get` methods defined on the `Pipeline`, similar to what we do in the `RunContext`. or maybe this is just an intermediate step and this will disappear in the future 🤷

File: cmd/skaffold/app/cmd/runner.go
Comment (yuwenma) [reply cmd/skaffold/app/cmd/runner.go]: Maybe not, providing a v2 validation is not a small change and may deserve another PR.

**Comment by yuwenma** [comment]
> I see a few `TODO` leftover in this PR, but I'm not sure they match up with the follow up work detailed in the description. if those aren't going to be addressed in this PR, can we leave a short note in the PR description for what needs to be done later?
> 
> I'm also feeling a bit strange about moving the type checking from the method signature into a simple typecast in the method - this seems brittle to me. as an example:
> 
> ```go
> func targetArtifacts(opts config.SkaffoldOptions, configs []util.VersionedConfig) []*latestV1.Artifact {
> 	var targetArtifacts []*latestV1.Artifact
> 	for _, cfg := range configs {
> 		for _, artifact := range cfg.(*latestV1.SkaffoldConfig).Build.Artifacts {
> ```
> 
> it might be better to abstract this away into `Get` methods defined on the `Pipeline`, similar to what we do in the `RunContext`. or maybe this is just an intermediate step and this will disappear in the future 🤷

The interface is my refactoring proposal, and this PR shows the implementation details how that could be done. The detail is in this [doc](https://docs.google.com/document/d/1GFy_5ya-ajjSVe1t6f7IWUlAQu35eTHAZZ1GkJBiqcA/edit?resourcekey=0-sKE-yXjuymQ0w2qqNXnTzw). Per offline discussion, I'll leave it to you to determine how the refactoring should work and I'm only going to focus on the branch v2.

**Review by tejal29 — APPROVED** [review]

**Comment by MarlonGamez** [comment]
There is one small linter error but I think this is pretty much ready to merge after that

**Pull request merged**
